oeps/architectural-decisions/oep-0065/decisions/0002-frontend-app-migrations.rst

```diff
@@ -0,0 +1,152 @@
+
+* ``dev`` will start a dev server, loading the repository's application modules into the shell in a site.
+* ``dev:module`` will start a dev server that provides the application modules via module federation.
+* ``build`` will create a standalone deployable artifact that uses the shell (similar to the micro-frontend architecture)
+* ``build:module`` will create a standalone deployable artifact that provides the application modules via module federation.
+* ``release`` will package the library for distribution on npm.
+* ``serve`` will work with ``build`` or ``build:module`` to locally serve the production assets they generated.
+* ``pack`` will work with ``release`` to create a ``.tgz`` file suitable for installing in local git checkouts that depend on the library. (this is a development tool)
+
+The ``dev``, ``dev:module``, ``build``, and ``build:module`` CLI commands will rely on the existence of a :term:`Site Config` file (the replacement for .env/env.config files) which will not be checked into the repository.
+
+Distributed as NPM Packages
+===========================
+
+``frontend-app-*`` repositories that are part of Open edX releases will be expected to be published on NPM as a library which exports its modules. These libraries will primarily be consumed by :term:`projects <Project>`.
+
+Consequences
+************
+
+As the module architecture stabilizes, ``frontend-app-*`` maintainers and developers will be encouraged to migrate their micro-frontends into application module libraries, and to adopt the module architecture provided by ``frontend-base``. (There will be a migration guide.)
+
+For micro-frontends that are migrated to application module libraries using the shell, there will be a deployment approach that mimics the micro-frontend architecture, but which will require operators to adopt a new underlying configuration and build process to achieve a similar result. Each ``frontend-app-*`` repository will need a deprecation process for the micro-frontend configuration and build infrastructure.
+
+Thinking in Modules
+===================
+
+Our definition of :term:`module` aligns with the `industry standard definition <https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules>`_. It is also used in the context of `module federation <https://module-federation.io>`_. It's a self-contained part of the frontend that represents a specific part of the :term:`Site`, and can be loaded in a variety of ways. We have several sub-types of module:
+
+* An *application module* represents a well-bounded sub-area of the Open edX frontend at a particular route path. This might be "courseware", "the login page", or "account settings". There are a number of application modules that are *required* for a functioning Open edX frontend Site.
```

Comment:
As per a previous review comment, I feel "application module" can double as the name for the whole thing (sans _library_ at the end).

Maybe another way to think about it is that different modules should live in different repos, lest we fall into the mono-repo issues we've discussed in the past.

Reply:
We talked about this offline, but for posterity - there are two related rationales for keeping modules (particularly application modules, which are generally pretty big) in the same repository together.  These are:

1. If they share code with other modules in the library.  If we split them up and have to create a secondary library to share code between them, we probably haven't made our lives better.  
2. If they're part of the same conceptual 'domain' of the site.  This was the original idea behind the MFE boundaries we have today, though we didn't execute it consistently.  Some are too big, some are too small.  It seems very reasonable to me that the learning-related stuff could be in "frontend-app-learning", and the user account/authentication related stuff would be in something like "frontend-app-user", which doesn't exist.